lib/Exporter/ONNXModelWriter.cpp

```diff
@@ -0,0 +1,1367 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Exporter/ONNXModelWriter.h"
+#include "glow/Graph/Utils.h"
+
+namespace glow {
+
+ONNXModelWriter::ONNXModelWriter(const std::string &modelFilename, Function &F,
+                                 size_t irVersion, size_t opsetVersion,
+                                 llvm::Error *errPtr, bool textMode)
+    : CommonOperatorWriter(modelFilename, F, errPtr),
+      opsetVersion_(opsetVersion) {
+  // if errPtr already contains an error then don't continue with constructor.
+  if (errPtr && *errPtr) {
+    return;
+  }
+
+  // Lambda to setup the ONNXModelWriter and return any llvm::Errors that were
+  // raised.
+  auto setup = [&]() -> llvm::Error {
+    // Loop through all nodes, output Graph to Model protobuf.
+    std::unordered_set<const Node *> reportedNodes;
+    ReporterType reporter = [&reportedNodes](const Node *node) {
+      reportedNodes.insert(node);
+    };
+    ONNX_TRAITS::ModelProto modelProto;
+    modelProto.set_ir_version(irVersion);
+    modelProto.set_producer_name("ONNXModelWriter");
+    auto *opsetImportProto = modelProto.add_opset_import();
+    opsetImportProto->set_version(opsetVersion);
+    auto *graphProto = modelProto.mutable_graph();
+    graphProto->set_name("glow");
+
+    // Use post order graph traversal.
+    // If node is constant with uses, turned into "input" and create a tensor
+    // If node is placeholder with uses, turned into "input" with tensor shape,
+    // except the case when placeholder has use as SaveNode.
+    // If node is on the ignore list - skip it.
+    // Otherwise call common operators method or special operators and write
+    // protobuf inputs from node inputs and protobuf outputs from uses.
+    GraphPostOrderVisitor visitor(G_);
+    for (const auto *N : visitor.getPostOrder()) {
+      if (isIgnoreNode(N)) {
+        // Skip ignored nodes.
+        continue;
+      }
+
+      if (reportedNodes.count(N)) {
+        continue;
+      }
+
+      const auto kind = N->getKind();
+      if (kind == Kinded::Kind::PlaceholderKind ||
+          kind == Kinded::Kind::ConstantKind) {
+        if (hasUsesOfKind(N, Kinded::Kind::SaveNodeKind)) {
+          // Storage as an input to SaveNode - ignore it.
+          continue;
+        }
+
+        // Handle placeholders cases.
+        if (kind == Kinded::Kind::PlaceholderKind) {
+          const auto *PH = llvm::dyn_cast<Placeholder>(N);
+          // Write global input, output only tensor shape.
+          auto *inputProto = graphProto->add_input();
+          tensorShapeFromPlaceholder(PH, inputProto);
+        } else {
+          // Write global initializer, output tensor bytes.
+          const auto *C = llvm::dyn_cast<Constant>(N);
```

Comment:
Since you do not expect this cast to fail you should use `llvm::cast`, which will assert if the value cannot be casted.